### PR TITLE
Unit testing best practices

### PR DESCRIPTION
**Pull request by dorschw**

### CODE REVIEW

File: docs/integrations/unit-testing.md [hunk 1]
```diff
@@ -225,6 +225,13 @@ def test_function():
 
 If the function raises a ValueError with proper error message, the test will pass.
 
+## Content Testing Guidelines
+
+* When testing a command, mock all external requests (e.g. mock `Client` methods), run the command on given inputs, and compare the output to a hardcoded value.
+* Prefer comparing whole outputs (e.g. comparing the output to a `CommandResult` object), to make sure the code behavior is preserved even when others add features and fix issues.
+* Use the `@freezegun.freeze_time` decorator on the testing method when the output may rely on the time of running.
```
Comment (tkatzir) [inline docs/integrations/unit-testing.md:232]:
```suggestion
* Use the `@freezegun.freeze_time` decorator on the testing method when the output may rely on the time or date (think Summer Time) of running.
```

**Review by tkatzir — COMMENTED** [review]
Looks good!

Line 238 should be replaced with:
script with `--console_log_threshold DEBUG` for verbose output.

File: docs/integrations/unit-testing.md (hunk 1 above)
Comment (dorschw) [reply docs/integrations/unit-testing.md:232]:
```suggestion
* Use the `@freezegun.freeze_time` decorator on the testing method when the output may vary in different time, date or timezone (e.g. daylight-saving time) in run time.
```

File: docs/integrations/unit-testing.md [hunk 2]
```diff
@@ -225,6 +225,13 @@ def test_function():
 
 If the function raises a ValueError with proper error message, the test will pass.
 
+## Content Testing Guidelines
+
+* When testing a command, mock all external requests (e.g. mock `Client` methods), run the command on given inputs, and compare the output to a hardcoded value.
+* Prefer comparing whole outputs (e.g. comparing the output to a `CommandResult` object), to make sure the code behavior is preserved even when others add features and fix issues.
+* Use the `@freezegun.freeze_time` decorator on the testing method when the output may vary in different time, date or timezone (e.g. daylight-saving time) in run time.
+* Methods that handle errors or exceptions should have multiple tests - one for the _happy path_ (normal functionality, without errors), and one testing method for each error handled (a.k.a _sad path/s_).
+
```
Comment (MosheEichler) [inline docs/integrations/unit-testing.md:234]:
```suggestion
* Don't include any sensitive data (secrets, tokens, email addresses, personal information, etc.).
```

Comment (dorschw) [reply docs/integrations/unit-testing.md:234]:
```suggestion
* Make sure to sanitize or remove any sensitive data (passwords, tokens, email addresses, personal information, etc.) from unit tests and test data.

```

File: docs/integrations/unit-testing.md [hunk 3]
```diff
@@ -225,6 +225,15 @@ def test_function():
 
 If the function raises a ValueError with proper error message, the test will pass.
 
+## Content Testing Guidelines
+
```
Comment (JudahSchwartz) [inline docs/integrations/unit-testing.md:229]: would add adise to mock demistomock directly using mocker, not rely on what we have in the file

File: docs/integrations/unit-testing.md [hunk 4]
```diff
@@ -225,6 +225,16 @@ def test_function():
...
+* When testing a command, mock all external requests (e.g. mock `Client` methods), run the command on given inputs, and compare the output to a hardcoded value.
```
Comment (JudahSchwartz) [inline docs/integrations/unit-testing.md:230]:
```suggestion
* When testing a command, mock all external requests (e.g. mock `Client` methods), run the command on given inputs, and compare the output to a hardcoded value.
* When using demistomock, mock demisto directly using *mocker*, don't rely on the defaults.
```

Comment (dorschw) [reply docs/integrations/unit-testing.md:230]:
```suggestion
* When testing a command, mock all external requests (e.g. mock `Client` methods), run the command on given inputs, and compare the output to a hardcoded value.
* When using demistomock, mock the `demisto` object directly using `mocker`, don't rely on the defaults.
```

**Pull request closed without merging**